hal/src/hal-sdcard.ads

```diff
@@ -0,0 +1,426 @@
+   SD_App_Get_CER_RN2                 : constant SD_Specific_Command := 46;
+   SD_App_Set_CER_RES2                : constant SD_Specific_Command := 47;
+   SD_App_Get_CER_RES1                : constant SD_Specific_Command := 48;
+   SD_App_Change_Secure_Area          : constant SD_Specific_Command := 49;
+   SD_App_Send_SCR                    : constant SD_Specific_Command := 51;
+
+   type Rsp_Kind is
+     (Rsp_Invalid,
+      --  Unknown/invalid command
+
+      Rsp_No,
+      --  No response.
+
+      Rsp_R1, Rsp_R1B, Rsp_R2, Rsp_R3, Rsp_R6, Rsp_R7
+      --  Response with format
+     );
+   --  Response format
+
+   type Tfr_Kind is
+     (Tfr_Invalid, Tfr_No,
+      Tfr_Read, Tfr_Write,
+      Tfr_Read_Multi, Tfr_Write_Multi);
+   --  Transfer
+
+   type Cmd_Desc_Type is record
+      Cmd : SD_Command;
+      Rsp : Rsp_Kind;
+      Tfr : Tfr_Kind;
+   end record;
+
+   type SDCard_Driver is limited interface;
+
+   procedure Reset
+     (This   : in out SDCard_Driver;
+      Status : out SD_Error) is abstract;
+   --  Initialize the driver, enable clocking.
+
+   procedure Set_Clock
+     (This : in out SDCard_Driver;
+      Freq : Natural) is abstract;
+   --  Set clock frequency.
+
+   procedure Set_Bus_Size
+     (This : in out SDCard_Driver;
+      Mode : Wide_Bus_Mode) is abstract;
+   --  Set host bus size; the command must have been set to the card.
+
+   procedure Send_Cmd
+     (This   : in out SDCard_Driver;
+      Cmd    : Cmd_Desc_Type;
+      Arg    : Unsigned_32;
+      Status : out SD_Error) is abstract;
+   --  Send a command (without data transfer) and wait for result.
+
+   procedure Send_Cmd
+     (This   : in out SDCard_Driver'Class;
+      Cmd    : SD_Command;
+      Arg    : Unsigned_32;
+      Status : out SD_Error);
+   --  Wrapper for Send_Cmd using a generic command.
+
+   procedure Send_ACmd
+     (This   : in out SDCard_Driver'Class;
+      Cmd    : SD_Specific_Command;
+      Rca    : Unsigned_16;
+      Arg    : Unsigned_32;
+      Status : out SD_Error);
+   --  Send application specific command
+
+   procedure Read_Cmd
+     (This   : in out SDCard_Driver;
+      Cmd    : Cmd_Desc_Type;
+      Arg    : Unsigned_32;
+      Buf    : System.Address;
+      Len    : Unsigned_32;
+      Status : out SD_Error) is abstract;
+   --  Read data command
+
+   procedure Read_Rsp48
+     (This : in out SDCard_Driver;
+      Rsp  : out Unsigned_32) is abstract;
+   --  Read the 32 interesting bits of the last 48bits response (start bit,
+   --  transmission bit, command index, crc and end bit are discarded).
+   --  Cannot fail.
+
+   procedure Read_Rsp136
+     (This           : in out SDCard_Driver;
+      W0, W1, W2, W3 : out Unsigned_32) is abstract;
+   --  Read the 128 interesting bits of the last 136 bit response.
+   --  W0 is the MSB, W3 the LSB
+   --  Cannot fail.
+
+   procedure Card_Identification_Process (Driver : in out SDCard_Driver'Class;
```

Comment:
> Gprbuild will certainly rightfully complain here about projects dependency circularity.

There's no circular dependency, services do not depend on drivers thanks to the HAL.

> Moreover we already have code in the HAL. 

Only in Bitmap, I plan to fix this someday.

> I don't see why the HAL should never have code: if this code still abstracts the hardware behaviour, to me that's OK.

These subprograms do not describe an interface or an abstraction, they use an interface/abstraction to compute data. They use the HAL to provide a service: the card identification.

> It would be limitating to prevent such common standard implementation there, or we should create a different layer for that, tightly linked to the HAL. 

I don't see any limitation so far, do you have an example?

> The service layer to me is very different, as it's above the drivers: it's pure software using services from the drivers to proceed. Filesystem is a good example, or TCP/IP.

What I suggest is in line with that.

HAL is at the heart of the library, it's seen by everything, it's used everywhere. The more code you put in there the more dependency you will have to deal with, that's when you will get in trouble. That why it's important to keep it as light weight and stand alone as possible.

In this specific instance there's no reason to cumber the HAL. As I said these subprograms do not describe an interface, they use an interface to provide a service.